falcon/routing/compiled.py

```diff
@@ -236,12 +256,24 @@ def insert(nodes, path_index=0):
             # exist and needs to be created. This builds a new branch of the
             # routing tree recursively until it reaches the new node leaf.
             new_node = CompiledRouterNode(path[path_index])
-            nodes.append(new_node)
+            if new_node.is_complex:
+                cpc = consume_path_converter(new_node)
+                if cpc:
+                    raise ValueError(
+                        'Cannot use converter "{1}" of variable "{0}" in a template '
+                        'that includes other characters or variables.'.format(*cpc)
+                    )
             if path_index == len(path) - 1:
+                nodes.append(new_node)
                 new_node.method_map = method_map
                 new_node.resource = resource
                 new_node.uri_template = uri_template
             else:
+                cpc = consume_path_converter(new_node)
+                if cpc:
+                    print('b')
+                    raise ValueError(no_children_err.format(uri_template, *cpc))
+                nodes.append(new_node)
```

Comment:
Why do we `nodes.append(new_node)` in both branches of the `if...else` as opposed to doing it once outside the statement?

Reply:
I guess the idea was to add it only once the checks are passed since raising an error after may lead to an inconsistent router state?

Reply:
good idea, will add a comment

Reply:
In the end it's probably better to just add the node and remove if before raising an error if needed